pkg/registry/common/expire/ns_server.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expire
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+)
+
+type nsServer struct {
+	server     registry.NetworkServiceRegistryServer
+	nseClient  registry.NetworkServiceEndpointRegistryClient
+	once       sync.Once
+	period     time.Duration
+	monitorErr error
+	nses       map[string]*registry.NetworkServiceEndpoint
+	nsCounter  map[string]int
+	nss        map[string]*registry.NetworkService
+	now        func() int64
+	sync.Mutex
+}
+
+func (n *nsServer) setGetTimeFunc(f func() int64) {
+	n.now = f
+}
+
+func (n *nsServer) setPeriod(d time.Duration) {
+	n.period = d
+}
+
+func (n *nsServer) monitor() {
+	go func() {
+		c, err := n.nseClient.Find(context.Background(), &registry.NetworkServiceEndpointQuery{
+			NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{},
+			Watch:                  true,
+		})
+		if err != nil {
+			n.monitorErr = err
+			return
+		}
+		for event := range registry.ReadNetworkServiceEndpointChannel(c) {
+			nse := event
+			n.Lock()
+			n.nses[nse.Name] = nse
+			for _, service := range nse.NetworkServiceNames {
```

Comment:
It looks not really correct, if for some of NetworkServices it would be a lot of updates, n.nsCounter for it will be big enough so service will be newer removed, since it is removed if and only if there is no network services. 
So I would recommend to handle a list of endpoints implementing a network service instead of having an counter.

Reply:
Good catch, I've added a check that incrementing should be only in case if nse does not exist in nses map.